lnbits/nodes/base.py

```diff
@@ -0,0 +1,239 @@
+
+        return cls(
+            counts=counts,
+            avg_size=int(
+                sum(channel.balance.total_msat for channel in channels) / len(channels)
+            ),
+            biggest_size=max(channel.balance.total_msat for channel in channels),
+            smallest_size=min(channel.balance.total_msat for channel in channels),
+            total_capacity=sum(channel.balance.total_msat for channel in channels),
+        )
+
+
+class NodeFees(BaseModel):
+    total_msat: int
+    daily_msat: Optional[int] = None
+    weekly_msat: Optional[int] = None
+    monthly_msat: Optional[int] = None
+
+
+class PublicNodeInfo(BaseModel):
+    id: str
+    backend_name: str
+    alias: str
+    color: str
+    num_peers: int
+    blockheight: int
+    channel_stats: ChannelStats
+    addresses: list[str]
+
+
+class NodeInfoResponse(PublicNodeInfo):
+    onchain_balance_sat: int
+    onchain_confirmed_sat: int
+    fees: NodeFees
+    balance_msat: int
+
+
+class NodePayment(BaseModel):
+    pending: bool
+    amount: int
+    fee: Optional[int] = None
+    memo: Optional[str] = None
+    time: int
+    bolt11: str
+    preimage: Optional[str]
+    payment_hash: str
+    expiry: Optional[float] = None
+    destination: Optional[NodePeerInfo] = None
+
+
+class NodeInvoice(BaseModel):
+    pending: bool
+    amount: int
+    memo: Optional[str]
+    bolt11: str
+    preimage: Optional[str]
+    payment_hash: str
+    paid_at: Optional[int] = None
+    expiry: Optional[int] = None
+
+
+class NodePaymentsResponse(BaseModel):
+    error_message: Optional[str]
+    payments: list[NodePayment]
+
+
+class NodeInvoiceFilters(FilterModel):
+    pass
+
+
+class NodePaymentsFilters(FilterModel):
+    pass
+
+
+class Node(ABC):
+    wallet: Wallet
+
+    def __init__(self, wallet: Wallet):
+        self.wallet = wallet
+        self.id: Optional[str] = None
+
+    @property
+    def name(self):
+        return self.__class__.__name__
+
+    async def get_id(self):
+        if not self.id:
+            self.id = await self._get_id()
+        return self.id
+
+    @abstractmethod
+    async def _get_id(self) -> str:
+        pass
+
+    async def get_peers(self) -> list[NodePeerInfo]:
+        ids = await self.get_peer_ids()
+        return [await self.get_peer_info(pubkey) for pubkey in ids]
+
+    @abstractmethod
+    async def get_peer_ids(self) -> list[str]:
+        pass
+
+    @abstractmethod
+    async def connect_peer(self, uri: str):
+        pass
+
+    @abstractmethod
+    async def disconnect_peer(self, id: str):
+        pass
+
+    @abstractmethod
+    async def _get_peer_info(self, pubkey: str) -> NodePeerInfo:
+        pass
+
+    async def get_peer_info(self, pubkey: str) -> NodePeerInfo:
+        info = cache.get(f"peers:{pubkey}")
+        if not info:
+            info = await self._get_peer_info(pubkey)
+            if info.last_timestamp:
+                cache.set(f"peers:{pubkey}", info)
```

Comment:
- the cache key might be too generic resulting in conflicts.
- something like `node,peers:{pubkey}` would be safer